taar/recommenders/similarity_recommender.py

```diff
@@ -0,0 +1,130 @@
+import logging
+from ..recommenders import utils
+from .base_recommender import BaseRecommender
+from scipy.spatial import distance
+
+s3_bucket = 'telemetry-private-analysis-2'
+DONOR_LIST_KEY = 'taar/legacy/addon_donors.json'
+SIM_TO_P_FUN_KEY = 'taar/legacy/test/lr_curves.json'
+
+logger = logging.getLogger(__name__)
+
+
+def fetch_donors_list_remote_resource():
+    return utils.get_s3_json_content(s3_bucket, DONOR_LIST_KEY)
+
+
+def fetch_lr_curve_remote_resource():
+    return utils.get_s3_json_content(s3_bucket, SIM_TO_P_FUN_KEY)
+
+
+class SimilarityRecommender(BaseRecommender):
+    """ A recommender class that returns top N addons based on the client similarity
+    with a set of candidate addon donors. Several telemetry fields are sued to compute
+    pairwise similarity with the donors and similarities are converted into a likelihood
+    ratio of being a good match versus not being a good match. These quantities are then
+    used to rank specific addons for recommendation.
+
+    This will load a json file containing updated list of addon donors updated periodically
+    by a separate weekly process using Longitdudinal Telemetry data.
+
+    This recommender may provide useful recommendations when collaborative_recommender
+    may not work.
+    """
+
+    def __init__(self):
+        # Download the addon donors list.
+        self.donors_list = fetch_donors_list_remote_resource()
+
+        if self.donors_list is None:
+            logger.error("Cannot download the JSON resource: {}".format(DONOR_LIST_KEY))
+
+        # Download the probability mapping curves from similarity to likelihood of being a good donor.
+        self.lr_curves = fetch_lr_curve_remote_resource()
+
+        if self.lr_curves is None:
+            logger.error("Cannot download the JSON resource: {}".format(SIM_TO_P_FUN_KEY))
+
+    def can_recommend(self, client_data):
+        # We can't recommend if we don't have our data files.
+        if self.donors_list is None or self.lr_curves is None:
+            return False
+
+        # Check that the client info contains a non-None value for each required telemetry field.
+        required_fields = ["geoCity", "subsessionLength", "locale", "os", "bookmark_count", "tab_open_count",
+                           "total_uri", "unique_tlds"]
+
+        for single_required_field in required_fields:
+            client_value = client_data.get(single_required_field, None)
+            if client_value is None:
+                logger.error("Malformed client data encountered in SimilarityRecommender, {} field not detected"
```

Comment:
nit: it's really "incomplete" more than "malformed", as clients might send partial data depending on when the ping is generated (e.g. `tab_open_count`).